### PR TITLE
Add fault injection macros and unit tests to rcl_lifecycle

### PR DESCRIPTION
**Pull request by brawner**
This adds fault injection macros and unit tests for rcl_lifecycle. This depends on #729. CI tests will come shortly.

Signed-off-by: Stephen Brawner <brawner@gmail.com>

### CODE REVIEW

**Comment by brawner** [comment]
Builds and tests of the following packages `rcutils`, `rcl_logging_spdlog`, `rosidl_typesupport_c`, `rosidl_typesupport_cpp`, `rcl`, `rcl_action`, `rcl_lifecycle`, `rosidl_runtime_c`, `rmw_dds_common`, `rmw`

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11916/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6961/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9709/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11850/)

**Comment by brawner** [comment]
This was rebased onto master after merging #727. It depends on https://github.com/ros2/rmw/pull/254. 

This tests this PR branch with current ros 2 branches and the `rmw` fault injection branch (`--packages-select rcl_lifecycle`). 
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12121/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7119/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9870/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12027/)

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp [hunk 1]
```diff
@@ -337,7 +340,6 @@ TEST(TestRclLifecycle, state_machine) {
   ret = rcl_lifecycle_state_machine_fini(&state_machine, nullptr, &allocator);
   EXPECT_EQ(ret, RCL_RET_ERROR);
   rcutils_reset_error();
-  std::cout << "state_machine: " << __LINE__ << std::endl;
```
Comment (Blast545) [inline rcl_lifecycle/test/test_rcl_lifecycle.cpp:340]: Should this be part of another PR to be able of backporting removing that msg?

File: rcl_lifecycle/test/test_default_state_machine.cpp [hunk 2]
```diff
@@ -833,3 +834,16 @@ TEST_F(TestDefaultStateMachine, default_sequence_error_unresolved) {
...
 }
+
+TEST_F(TestDefaultStateMachine, init_fini_maybe_fail) {
+  RCUTILS_FAULT_INJECTION_TEST(
+  {
+    // If this in zero initialized outside of this loop, there is potential for a segfault
+    // when attempting to init a second time
+    rcl_lifecycle_state_machine_t sm = rcl_lifecycle_get_zero_initialized_state_machine();
+    rcl_ret_t ret = rcl_lifecycle_init_default_state_machine(&sm, this->allocator);
+    if (RCL_RET_OK == ret) {
```
Comment (Blast545) [inline rcl_lifecycle/test/test_default_state_machine.cpp:845]: reset error if init fails?

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp (hunk 1 above)
Comment (brawner) [reply rcl_lifecycle/test/test_rcl_lifecycle.cpp:340]: Sounds good. Opened https://github.com/ros2/rcl/pull/773

File: rcl_lifecycle/test/test_default_state_machine.cpp (hunk 2 above)
Comment (brawner) [reply rcl_lifecycle/test/test_default_state_machine.cpp:845]: Done

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp (hunk 1 above)
Comment (brawner) [reply rcl_lifecycle/test/test_rcl_lifecycle.cpp]: Rebased onto #773

**Comment by brawner** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12128/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7123/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9875/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12031/)

**Comment by brawner** [comment]
Rebasing back on to master after merging #773
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12134/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7128/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9879/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12036/)

File: rcl_lifecycle/test/test_default_state_machine.cpp [hunk 3]
```diff
@@ -833,3 +834,21 @@ TEST_F(TestDefaultStateMachine, default_sequence_error_unresolved) {
       rcl_lifecycle_state_machine_fini(&state_machine, this->node_ptr, this->allocator));
   }
 }
+
+TEST_F(TestDefaultStateMachine, init_fini_maybe_fail) {
+  RCUTILS_FAULT_INJECTION_TEST(
+  {
+    // If this in zero initialized outside of this loop, there is potential for a segfault
+    // when attempting to init a second time
```
Comment (hidmic) [inline rcl_lifecycle/test/test_default_state_machine.cpp:842]: @brawner hmm, shouldn't `fini` take care of leaving `rcl_lifecycle_state_machine_t` in a known state?

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp [hunk 4]
```diff
@@ -436,3 +439,50 @@ TEST(TestRclLifecycle, state_transitions) {
...
+  rcl_ret_t ret = rcl_init_options_init(&init_options, allocator);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+  ret = rcl_init(0, nullptr, &init_options, &context);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    ASSERT_EQ(RCL_RET_OK, rcl_shutdown(&context));
+    ASSERT_EQ(RCL_RET_OK, rcl_context_fini(&context));
```
Comment (hidmic) [inline rcl_lifecycle/test/test_rcl_lifecycle.cpp:458]: @brawner for this two, why not `EXPECT_EQ` like elsewhere?

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp [hunk 5]
```diff
@@ -436,3 +439,50 @@ TEST(TestRclLifecycle, state_transitions) {
...
+
+  RCUTILS_FAULT_INJECTION_TEST(
+  {
+    // Init segfaults if this is not zero initialized
+    rcl_lifecycle_state_machine_t sm = rcl_lifecycle_get_zero_initialized_state_machine();
+
+    ret = rcl_lifecycle_state_machine_init(
+      &sm, &node, pn, cs, gs, gas, gat, gtg, true, &allocator);
+    if (RCL_RET_OK == ret) {
+      ret = rcl_lifecycle_state_machine_fini(&sm, &node, &allocator);
```
Comment (hidmic) [inline rcl_lifecycle/test/test_rcl_lifecycle.cpp:485]: @brawner can `rcl_lifecycle_state_machine_fini` fail due to fault injection?

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp (hunk 4 above)
Comment (brawner) [reply rcl_lifecycle/test/test_rcl_lifecycle.cpp]: Switched these to EXPECT_EQ, here and below.

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp (hunk 5 above)
Comment (brawner) [reply rcl_lifecycle/test/test_rcl_lifecycle.cpp:485]: Adding a second finalization if this one fails.

File: rcl_lifecycle/test/test_default_state_machine.cpp (hunk 3 above)
Comment (brawner) [reply rcl_lifecycle/test/test_default_state_machine.cpp]: Looks like this was the case before I finalized a second time. This comment is no longer accurate. Moving zero_initialization out of this test.

**Comment by brawner** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12156/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7144/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9890/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12046/)

**Comment by hidmic** [comment]
ARM64 CI had a fluke. Updated the badge link.
